### PR TITLE
Implements User Dashboard front-end

### PR DESCRIPTION
**Pull request by qhuynh96**
* A basic feed of User's open PRs.
* Desktop view:
user-dashboard
user-dashboard-2
* Mobile view:
user-dashboard-mobile

### CODE REVIEW

**Comment by stevekuznetsov** [comment]
Is the `size/XL` label in the `to remove` column simply an artifact of test data that you were using?

**Comment by qhuynh96** [comment]
Test data :p

File: prow/cmd/deck/static/user-script.js [hunk 1]
```diff
@@ -0,0 +1,500 @@
...
+
+    const jobIndicator = document.createElement("SPAN");
+    const js = jobStatus(builds);
+    jobIndicator.classList.add("indicator", js);
+    jobIndicator.title = js[0].toUpperCase() + js.slice(1) + " job(s)";
+    const mergeIndicator = document.createElement("SPAN");
+    const canMerge = isAbleToMerge(queries);
+    mergeIndicator.classList.add("indicator",
+        canMerge ? "successful" : "failed");
+    mergeIndicator.title = canMerge ? "Meet merge requirement"
```
[the remark this answers is not in the document]
Comment (qhuynh96) [reply prow/cmd/deck/static/user-script.js]: done

File: prow/cmd/deck/static/user-script.js [hunk 2]
```diff
@@ -0,0 +1,500 @@
...
+    const status = document.createElement("DIV");
+    const failedJobs = builds.filter(build => {
+        return build.state === "failure";
+    });
+    // Job status indicator
+    const state = jobStatus(builds);
+    let statusText = "";
+    switch (state) {
+        case "successful":
+            statusText = "All jobs succeed";
```
[the remark this answers is not in the document]
Comment (qhuynh96) [reply prow/cmd/deck/static/user-script.js]: done

File: prow/cmd/deck/static/user-script.js [hunk 3]
```diff
@@ -0,0 +1,500 @@
...
+    table.classList.add("merge-table");
+    if (isAbleToMerge(queries)) {
+        table.classList.add("hidden");
+    }
+    const thead = document.createElement("THEAD");
+    const tableRow = document.createElement("TR");
+    const col1 = document.createElement("TD");
+    col1.innerHTML = "To have";
+    const col2 = document.createElement("TD");
+    col2.innerHTML = "To remove";
```
[the remark this answers is not in the document]
Comment (qhuynh96) [reply prow/cmd/deck/static/user-script.js]: did you mean about the variable name or "to remove" term?

File: prow/cmd/deck/static/user-script.js [hunk 4]
```diff
@@ -0,0 +1,500 @@
...
+
+/**
+ * Returns an icon element.
+ * @param {string} iconString icon name
+ * @param {string} tooltip tooltip string
+ * @return {Element}
+ */
+function createIcon(iconString, tooltip = "") {
+    const icon = document.createElement("I");
+    icon.classList.add(...["icon-button", "material-icons"]);
```
[the remark this answers is not in the document]
Comment (qhuynh96) [reply prow/cmd/deck/static/user-script.js]: done

File: prow/cmd/deck/static/user-script.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (stevekuznetsov) [reply prow/cmd/deck/static/user-script.js]: The user-facing "to have" and "to remove" phrases. Some ideas:

| Missing Labels | Forbidden Labels | 
| - | - |
| `lgtm` | `do-not-merge` |

I don't like `"forbidden"` too much either though? IDK.

File: prow/cmd/deck/static/user-script.js [hunk 5]
```diff
@@ -0,0 +1,500 @@
...
+ * @param builds
+ * @return {boolean}
+ */
+function jobStatus(builds) {
+    if (builds.length === 0) {
+        return "successful";
+    }
+    switch (builds[0].state) {
+        case "success":
+            return "successful";
```
[the remark this answers is not in the document]
Comment (qhuynh96) [reply prow/cmd/deck/static/user-script.js]: done

File: prow/cmd/deck/static/user-script.js [hunk 6]
```diff
@@ -0,0 +1,500 @@
...
+    const button = document.createElement("BUTTON");
+    button.classList.add("mdl-button", "mdl-js-button", "mdl-button--raised",
+        "mdl-button--primary", "mdl-js-ripple-effect");
+    button.innerHTML = "Login to Github";
+    button.addEventListener("click", () => {
+        const url = window.location;
+        window.location.href = url.origin + "/user-dashboard/login";
+    });
+    const intro = document.createElement("H3");
+    intro.innerHTML = "User Dashboard needs you login to your Github account";
```
[the remark this answers is not in the document]
Comment (qhuynh96) [reply prow/cmd/deck/static/user-script.js]: done

File: prow/cmd/deck/static/user-script.js [hunk 7]
```diff
@@ -0,0 +1,500 @@
...
+    return cardBody;
+}
+
+/**
+ * Compare function that prioritizes jobs which are in failure state.
+ * @param a
+ * @param b
+ * @return {number}
+ */
+function compareJobFn(a, b) {
```
[the remark this answers is not in the document]
Comment (qhuynh96) [reply prow/cmd/deck/static/user-script.js]: sorry dont get this comment

File: prow/cmd/deck/static/user-script.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (qhuynh96) [reply prow/cmd/deck/static/user-script.js]: Good terms, they may be opposite to payload data (we currently using labels and missingLabels), but for users these terms are clear enough

File: prow/cmd/deck/static/user-script.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (stevekuznetsov) [reply prow/cmd/deck/static/user-script.js]: Just saying I like the way you sorted the jobs here -- it makes a lot of sense!

[the remark this answers is not in the document]
Comment (qhuynh96) [reply prow/cmd/deck/static/user-script.js]: oh thanks, though something need to be resolved.

File: prow/cmd/deck/static/user-script.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (qhuynh96) [reply prow/cmd/deck/static/user-script.js]: How about "Blocking Labels"

**Comment by BenTheElder** [comment]
/area prow
/assign
Discussed some UX nits offline, but definitely excited for this! Will take a pass once after nits are updated.

**Comment by qhuynh96** [comment]
@stevekuznetsov I made some changes to refine the UI. You may want to review (changes) again.
@BenTheElder From our offline discusses, I tried to put the labels to the bottom of the card, but it does not look well as the labels seem not relevant / link well to the rest of the card (we only need to see our labels when we want to examine queries) so I put it to the top of the table as also discussed. WDYT?

About the label behavior, I cleared the opacity, so all of them should be displayed with 100% opacity now. However, some labels that we don't care are hidden by default, we can always see them by toggling the information button. Probably move them to a dialog (if necessary) tmr.

File: prow/cmd/deck/static/user-script.js [hunk 8]
```diff
@@ -0,0 +1,556 @@
...
+ * requirements, it will show that the PR needs to resolve labels.
+ * @param jobStatus
+ * @param canMerge
+ * @return {Element}
+ */
+function createTitleLabel(jobStatus, canMerge) {
+    const label = document.createElement("SPAN");
+    label.classList.add("title-label", "mdl-shadow--2dp");
+    if (!canMerge) {
+        label.innerHTML = "Needs to Resolve Labels";
```
Comment (BenTheElder) [inline prow/cmd/deck/static/user-script.js]: How about "Incorrect Labels for Merge" ?

**Comment by BenTheElder** [comment]
@qhuynh96 I think you should take a look at this with some more real PR data next, this is shaping up to be very shiny though 😄 
/hold
^ data :P

**Comment by qhuynh96** [comment]
Should I create dummy PRs for testing? Or I can get setup to get your PRs and test? WDYT?

**Comment by BenTheElder** [comment]
Either and/or both works

File: prow/cmd/deck/static/user-script.js [hunk 9]
```diff
@@ -0,0 +1,556 @@
...
+    if (canMerge) {
+        status.innerHTML = "Meets merge requirements";
+        status.classList.add("status", "succeeded");
+    } else {
+        status.innerHTML = "Does not meet merge requirements";
+        status.classList.add("status", "failed");
+    }
+    if (!canMerge) {
+        const iconButton = createIcon("help",
+            "To meet merge requirements, the PR needs to get all required labels and resolve all forbidden labels for one row at least.",
```
Comment (stevekuznetsov) [inline prow/cmd/deck/static/user-script.js]: There may be clearer wording for this. When do we have multiple rows?

File: prow/cmd/deck/static/user-script.js [hunk 10]
```diff
@@ -0,0 +1,556 @@
...
+    const button = document.createElement("BUTTON");
+    button.classList.add("mdl-button", "mdl-js-button", "mdl-button--raised",
+        "mdl-button--primary", "mdl-js-ripple-effect");
+    button.innerHTML = "Login to Github";
+    button.addEventListener("click", () => {
+        const url = window.location;
+        window.location.href = url.origin + "/user-dashboard/login";
+    });
+    const intro = document.createElement("H3");
+    intro.innerHTML = "User Dashboard needs you login to grant it OAuth scopes";
```
Comment (stevekuznetsov) [inline prow/cmd/deck/static/user-script.js]:
```js
    intro.innerHTML = "User Dashboard needs you to login and grant it OAuth scopes"; 
```

 ?

**Comment by BenTheElder** [comment]
Discussed this some more offline, I think we should generalize this to a "PR Status" page that **defaults** to _your_ PRs, but has some filtering at the top that can be pointed at repos / PR numbers.

If we do this then we can also link to this page at a specific PR from the tide status context's link which will should give users a nicer overview of the PR state and help them discover that this dashboard exists. We could have an quick "reset" button on the filtering to take them to their own PRs.

**Comment by qhuynh96** [comment]
Another update: 
So today I was working on UI/UX feedback from everyone (mostly from @BenTheElder, thanks Ben!). Overall, I feel like it reaches the stage that can be checked in. I am aware of existing issues with the current version but I would prefer to address those in the follow up as I have some demo sessions that should be delivered by next week. WDYT? @BenTheElder @stevekuznetsov @cjwagner

**Comment by stevekuznetsov** [comment]
Like the idea to make this a generic PR status viewer that defaults to _your_ PRs but also agreed that that can come in a follow-up.

**Comment by BenTheElder** [comment]
Agreed. Let's push forward with the first ~1k lines of this we have now and come back for another round :-)

**Comment by qhuynh96** [comment]
Great, so anyone wants to do the last review or give it an approval

**Comment by BenTheElder** [comment]
Reading through now :-)

**Comment by stevekuznetsov** [comment]
Read looked good but @BenTheElder can push the button

**Comment by BenTheElder** [comment]
Side note: we may want to make it less effort to work on this with runlocal in the future, which may be tricky. :shrug:

**Pull request merged**